src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Managed.cs

```diff
@@ -42,118 +42,80 @@ public void Abort()
-                        handler.SslOptions.ClientCertificates = new X509Certificate2Collection();
-                        handler.SslOptions.ClientCertificates.AddRange(options.ClientCertificates);
-                    }
-                }
+                        // Issue the request.  The response must be status code 101.
+                        CancellationTokenSource? linkedCancellation;
+                        CancellationTokenSource externalAndAbortCancellation;
+                        if (cancellationToken.CanBeCanceled) // avoid allocating linked source if external token is not cancelable
+                        {
+                            linkedCancellation =
+                                externalAndAbortCancellation =
+                                CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, _abortSource.Token);
+                        }
+                        else
+                        {
+                            linkedCancellation = null;
+                            externalAndAbortCancellation = _abortSource;
+                        }
 
-                // Issue the request.  The response must be status code 101.
-                CancellationTokenSource? linkedCancellation;
-                CancellationTokenSource externalAndAbortCancellation;
-                if (cancellationToken.CanBeCanceled) // avoid allocating linked source if external token is not cancelable
-                {
-                    linkedCancellation =
-                        externalAndAbortCancellation =
-                        CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, _abortSource.Token);
-                }
-                else
-                {
-                    linkedCancellation = null;
-                    externalAndAbortCancellation = _abortSource;
-                }
+                        using (linkedCancellation)
+                        {
+                            response = await invoker.SendAsync(request, externalAndAbortCancellation.Token).ConfigureAwait(false);
+                            externalAndAbortCancellation.Token.ThrowIfCancellationRequested(); // poll in case sends/receives in request/response didn't observe cancellation
+                        }
 
-                using (linkedCancellation)
-                {
-                    response = await new HttpMessageInvoker(handler).SendAsync(request, externalAndAbortCancellation.Token).ConfigureAwait(false);
-                    externalAndAbortCancellation.Token.ThrowIfCancellationRequested(); // poll in case sends/receives in request/response didn't observe cancellation
-                }
+                        ValidateResponse(response, secValue, options);
+                        break;
+                    }
+                    catch (HttpRequestException ex) when
+                        (ex.Data.Contains("SETTINGS_ENABLE_CONNECT_PROTOCOL") &&
```

Comment:
Please correct me if I'm wrong, but the exception seems to contain SETTINGS_ENABLE_CONNECT_PROTOCOL only when H2 connection is established but settings did not contain enabling extended connect? What will happen if H2 connection was not established at all (if server is H1 only)?

Also, I am a bit concerned about the whole `while(true)` approach. The conditions on when we exit or continue looping are a bit obscured.
We only have 3 options: try only H2, try H2 and then H1, try only H1. Meaning, we only have 2 steps, and the second will only be called if the first was not allowed or not succeeded in a special way. Can we unwind the loop? Something like
```
res = null
fail if unexpected versions
if (can try H2)
{
    try
    {
        res = do H2 stuff
    }
    catch when can downgrade
}
if (res == null) // either H2 not requested or it failed with downgrade enabled
{
    res = do H1 stuff
}
do stuff with res
```
Please let me know if I miss something

Reply:
> What will happen if H2 connection was not established at all (if server is H1 only)?

In that case we get `HttpRequestException` that I cannot separate from other reasons. I'm not sure if I can add `Data` to the exception that is already in use?